# 1. Mean Absolute Error

##### MAE is the average of differences between actual and predicted values and it calculates on avg how much your predictions are wrong
##### Lower MAE = Better model

In [20]:
from sklearn.metrics import mean_absolute_error as mae
actual = [2, 3, 6, 7, 8]
calculated = [3, 3, 7, 9, 11]
error= mae(actual, calculated)
print("MAE:", error)

MAE: 1.4


##### Here my model is wrong by 1.4 units on average

# 2. Mean Squared Error

##### MSE is the average of squared differences between actual and predicted values. It calculates on squared avg how much your prediction are wrong

In [23]:
from sklearn.metrics import mean_squared_error as mse
actual = [2, 3, 6, 7, 8]
calculated = [3, 3, 7, 9, 11]
error= mse(actual, calculated)
print("MSE:",error)

MSE: 3.0


# 3. Root Mean Squared Error

##### RMSE is the root of MSE.

In [24]:
from sklearn.metrics import mean_squared_error as mse
import numpy as np
actual = [2, 3, 6, 7, 8]
calculated = [3, 3, 7, 9, 11]
error= mse(actual, calculated)
rmse= np.sqrt(error)
print("RMSE:",rmse)

RMSE: 1.7320508075688772


# 4. R2 Score

##### R2 score tells you how well your regression model fits the data.
##### if value= 0 then, errors of regression line and mean line is same.
##### if value= 1 then it is a perfect model, regression line is not doing any mistakes.
##### if value< 0 then, regression line prediction is bad infact they are worse than predicting mean.

In [29]:
from sklearn.metrics import r2_score as r2s
actual = [2, 3, 6, 7, 8]
calculated = [3, 3, 7, 9, 11]
r2score= r2s(actual, calculated)
print("R2 score:", r2score)

R2 score: 0.4402985074626866


##### About 44% of the variation in actual values is explained by the model.
##### Remaining 56% is unexplained error.

In [31]:
from sklearn.metrics import r2_score as r2s
actual = [2, 3, 6, 7, 8]
calculated = [2, 3, 6, 7, 9]
r2score= r2s(actual, calculated)
print("R2 score:", r2score)

R2 score: 0.9626865671641791


##### The model has only 0.04% error

In [32]:
from sklearn.metrics import r2_score as r2s
actual = [2, 3, 6, 7, 8]
calculated = [2, 3, 6, 7, 8]
r2score= r2s(actual, calculated)
print("R2 score:", r2score)

R2 score: 1.0


##### Here we can see the model has perfectly predicted the score

# 5.Adjusted R2 score

##### Adjusted R² is a modified version of R² that accounts for the number of features (independent variables) in your model.

In [59]:
from sklearn.metrics import r2_score as r2s
actual = [2, 3, 6, 7, 8]
calculated = [2, 4, 7, 8, 8]
r2score= r2s(actual, calculated)
print("R2 score:", r2score)
n= len(actual)  #number of data
k= 1 #number of independent variable(input)
adjr2= 1- ((1-r2score) * (n-1))/(n-1-k)
print("Adjusted R2 score:", adjr2)

R2 score: 0.8880597014925373
Adjusted R2 score: 0.8507462686567164


##### R² = 0.888 → model explains 88.8% of variation
##### Adjusted R² = 0.851 → slightly lower because we only have 1 feature, but formula accounts for model complexity

### Adjusted R2 score with two input

In [60]:
from sklearn.metrics import r2_score as r2s
from sklearn.linear_model import LinearRegression
import numpy as np
# Two input features
# X1 = Hours studied
# X2 = Sleep hours

X = np.array([
    [2, 6],
    [3, 7],
    [4, 8],
    [5, 8],
    [6, 9]
])

# Actual output (Marks)
y = np.array([50, 55, 65, 70, 75])

lr= LinearRegression()
lr.fit(X, y)

pred= lr.predict(X)
print("Actual Score:", y)
print("Predicted Score:", pred)

r2score= r2s(y, pred)
print("R2 Score:", r2score)

n= len(X)
k=2
adjr2= 1- ((1-r2score) * (n-1))/(n-1-k)
print("Adjusted R2 score:", adjr2)

Actual Score: [50 55 65 70 75]
Predicted Score: [49.66666667 56.66666667 63.66666667 69.         76.        ]
R2 Score: 0.9844961240310078
Adjusted R2 score: 0.9689922480620157
